src/webgpu/util/conversion.ts

```diff
@@ -101,11 +101,66 @@ export function float32ToFloat16Bits(n: number) {
  * Decodes an IEEE754 16 bit floating point number into a JS `number` and returns.
  */
 export function float16BitsToFloat32(float16Bits: number): number {
-  const buf = new DataView(new ArrayBuffer(Float32Array.BYTES_PER_ELEMENT));
-  // shift exponent and mantissa bits and fill with 0 on right, shift sign bit
-  buf.setUint32(0, ((float16Bits & 0x7fff) << 13) | ((float16Bits & 0x8000) << 16), true);
-  // shifting for bias different: f16 uses a bias of 15, f32 uses a bias of 127
-  return buf.getFloat32(0, true) * 2 ** (127 - 15);
+  return floatBitsToNumber(float16Bits, kFloat16Format);
+}
+
+type FloatFormat = { signed: 0 | 1; exponentBits: number; mantissaBits: number; bias: number };
+
+/** FloatFormat defining IEEE754 32-bit float. */
+export const kFloat32Format = { signed: 1, exponentBits: 8, mantissaBits: 23, bias: 127 } as const;
+/** FloatFormat defining IEEE754 16-bit float. */
+export const kFloat16Format = { signed: 1, exponentBits: 5, mantissaBits: 10, bias: 15 } as const;
+
+/**
+ * Decodes an IEEE754 float with the supplied format specification into a JS number.
+ *
+ * The format MUST be no larger than a 32-bit float.
+ */
+export function floatBitsToNumber(bits: number, fmt: FloatFormat): number {
+  const signBit = bits >>> (fmt.exponentBits + fmt.mantissaBits);
+  const exptBits = (bits >>> fmt.mantissaBits) & ((1 << fmt.exponentBits) - 1);
+  const mantBits = bits & ((1 << fmt.mantissaBits) - 1);
+
+  // Pad the provided bits out to f32, then convert to a `number` with the wrong bias.
+  // E.g. for f16 to f32:
+  // - f16: S    EEEEE MMMMMMMMMM
+  //        ^ 000^^^^^ ^^^^^^^^^^0000000000000
+  // - f32: S eeeEEEEE MMMMMMMMMMmmmmmmmmmmmmm
+  let f32BitsWithWrongBias = signBit;
+  f32BitsWithWrongBias <<= kFloat32Format.exponentBits;
+  f32BitsWithWrongBias |= exptBits;
+  f32BitsWithWrongBias <<= fmt.mantissaBits;
+  f32BitsWithWrongBias |= mantBits;
+  f32BitsWithWrongBias <<= kFloat32Format.mantissaBits - fmt.mantissaBits;
+  const numberWithWrongBias = new Float32Array(new Uint32Array([f32BitsWithWrongBias]).buffer)[0];
+  return numberWithWrongBias * 2 ** (kFloat32Format.bias - fmt.bias);
+}
+
+/**
+ * Given a floating point number (as an integer representing its bits), computes how many ULPs it is
+ * from zero.
+ *
+ * Subnormal numbers are skipped, so that 0 is one ULP from the minimum normal number.
+ * Subnormal values are flushed to 0.
+ * Positive and negative 0 are both considered to be 0 ULPs from 0.
+ */
+export function floatBitsToNormalULPFromZero(bits: number, fmt: FloatFormat): number {
```

Comment:
For copy_to_texture 16-bit float(and 32-bit float) result comparasion. With this helper function, it seems that we could check the result by:
```
// Assume expect is larger than actual
floatBitsToNormalULPFromZero(Uint16(expected) - Uint16(actual), kFloat16Format) < constant?
```
Am I right?

Reply:
If this is the direction, I'm a bit worry about the case running time.
As you may know, current compare logic is [simple and hack](https://github.com/gpuweb/cts/pull/1043/files#diff-0781362fbbe96f1c4922e18e4ccead52e6c15573ed9bfa091c5b039c940e4d08R268)
But it still took longer time because it requires a [buffer view reinterpretation](https://github.com/gpuweb/cts/pull/1043/files#diff-0781362fbbe96f1c4922e18e4ccead52e6c15573ed9bfa091c5b039c940e4d08R246).
If we add an extra ops `floatBitsToNormalULPFromZero`, I think it took longer time.

So maybe a bit hack but do you think it is possible that we took everything as Uint8 as input and do the bit ops? It will save the buffer view reinterpretation.

Reply:
And another option is to save time on the other place rather than the float compare.

Reply:
Performance is definitely a potential issue and I haven't investigated it enough yet, thanks for highlighting it.
It's even worse than your example code, because it's more like `floatBitsToNormalULPFromZero(expected) - floatBitsToNormalULPFromZero(actual)`.

We have a `diffULP` helper already for directly determining the ULPs between two values without computing them relative to zero. I'll investigate the performance and see what can be done.

Reply:
As a point of comparison,
`webgpu:web_platform,copyToTexture,ImageBitmap:from_ImageData:alpha="none";orientation="none";srcDoFlipYDuringCopy=true;dstColorFormat="rgba16float";*`
before: 2500ms each
after: 4300ms each

Not quite as bad as I expected, but could probably be better

Reply:
> We have a diffULP helper already for directly determining the ULPs between two values without computing them relative to zero.

Reviewing, I don't think this would help. It's doing almost the same stuff.

> But it still took longer time because it requires a [buffer view reinterpretation](https://github.com/gpuweb/cts/pull/1043/files#diff-0781362fbbe96f1c4922e18e4ccead52e6c15573ed9bfa091c5b039c940e4d08R246).

And reinterpretation would still be happening if I used diffULP. That said, I wouldn't expect reinterpretation to be slow, so not sure yet where exactly the slowdown comes from.

Reply:
Looking at https://github.com/gpuweb/cts/pull/1043/commits/74816819e165625a094f6060d8134281f21ecb46, I'm guessing it was `float16BitsToFloat32`/`float16BitsToFloat32`. Which is probably a little more expensive than `floatBitsToNormalULPFromZero` though I wouldn't expect it to be that much worse.

Reply:
>Looking at https://github.com/gpuweb/cts/commit/74816819e165625a094f6060d8134281f21ecb46, I'm guessing it was float16BitsToFloat32/float16BitsToFloat32. Which is probably a little more expensive than floatBitsToNormalULPFromZero though I wouldn't expect it to be that much worse.

Yes, removing these two helper functions help accelerated the tests a lot but it is still worse than the [Uint8 comparation](https://github.com/gpuweb/cts/commit/74816819e165625a094f6060d8134281f21ecb46#diff-0781362fbbe96f1c4922e18e4ccead52e6c15573ed9bfa091c5b039c940e4d08L321) a lot (on my machine) but the same performance as [float32 comparation](https://github.com/gpuweb/cts/commit/74816819e165625a094f6060d8134281f21ecb46#diff-0781362fbbe96f1c4922e18e4ccead52e6c15573ed9bfa091c5b039c940e4d08L309). So I suspect this is due to the reinterpretation (But I think it shouldn't took long time).

 >webgpu:web_platform,copyToTexture,ImageBitmap:from_ImageData:alpha="none";orientation="none";srcDoFlipYDuringCopy=true;dstColorFormat="rgba16float";*
before: 2500ms each
after: 4300ms each

Thanks for testing! I understand that 4300ms is the time that applying `diffULP`, right?

Reply:
I dug down into the performance of the ImageBitmap:from_ImageData test and found that it was a simple matter of implementing this optimization I had left for myself (in #1055):
```ts
    // MAINTENANCE_TODO: Could be faster to actually implement numberToBits directly.
    numberToBits: (components: PerTexelComponent<number>) =>
      ret.unpackBits(new Uint8Array(ret.pack(encode(components)))),
```
before: 2500ms
draft: 4300ms
after: 2030ms!

Reply:
For another test case:
https://gpuweb.github.io/cts/standalone/?runnow=1&worker=0&debug=1&q=webgpu:web_platform,copyToTexture,canvas:copy_contents_from_gpu_context_canvas:canvasType=%22onscreen%22;srcAndDstInSameGPUDevice=true;dstColorFormat=%22rgba16float%22;srcPremultiplied=true;dstPremultiplied=true;srcDoFlipYDuringCopy=true

before: 1490ms
after: 1410ms

Reply:
Cool!